controller/db/controller_setting_store.go

```diff
@@ -0,0 +1,375 @@
+	})
+
+	store.AddListener(store.onDelete, boltz.EntityDeleted)
+}
+
+type untypedConstraintWithCtx[E boltz.Entity] struct {
+	eventListener func(boltz.MutateContext, boltz.Entity)
+	changeTypes   []boltz.EntityEventType
+}
+
+func (self *untypedConstraintWithCtx[E]) ProcessPreCommit(*boltz.EntityChangeState[E]) error {
+	return nil
+}
+
+func (self *untypedConstraintWithCtx[E]) ProcessPostCommit(state *boltz.EntityChangeState[E]) {
+	for _, changeType := range self.changeTypes {
+		if state.ChangeType == boltz.EntityCreated && changeType.IsCreate() {
+			if changeType.IsAsync() {
+				go self.eventListener(state.Ctx, state.FinalState)
+			} else {
+				self.eventListener(state.Ctx, state.FinalState)
+			}
+		} else if state.ChangeType == boltz.EntityUpdated && changeType.IsUpdate() {
+			if changeType.IsAsync() {
+				go self.eventListener(state.Ctx, state.FinalState)
+			} else {
+				self.eventListener(state.Ctx, state.FinalState)
+			}
+		} else if state.ChangeType == boltz.EntityDeleted && changeType.IsDelete() {
+			if changeType.IsAsync() {
+				go self.eventListener(state.Ctx, state.InitialState)
+			} else {
+				self.eventListener(state.Ctx, state.InitialState)
+			}
+		}
+	}
+}
+
+func (store *controllerSettingStoreImpl) initializeLinked() {}
+
+func (store *controllerSettingStoreImpl) NewEntity() *ControllerSetting {
+	return &ControllerSetting{
+		Oidc: &OidcSettingDef{
+			RedirectUris:   nil,
+			PostLogoutUris: nil,
+		},
+	}
+}
+
+func (store *controllerSettingStoreImpl) DeleteById(ctx boltz.MutateContext, id string) error {
+	policy, err := store.LoadById(ctx.Tx(), id)
+	if err != nil {
+		return err
+	}
+	if !policy.IsSystem || ctx.IsSystemContext() {
+		return store.BaseStore.DeleteById(ctx, id)
+	}
+	return errorz.NewEntityCanNotBeDeletedFrom(errors.New("default settings cannot be removed, only updated"))
+}
+
+type ControllerSettingsEvent struct {
+	Effective *ControllerSetting
+	Instance  *ControllerSetting
+	Global    *ControllerSetting
+}
+
+type ControllerSettingListener func(setting string, controllerId string, settingEvent *ControllerSettingsEvent)
+
+func (store *controllerSettingStoreImpl) Watch(setting string, listener ControllerSettingListener, controllerId string, additionalControllerIds ...string) {
+
+	additionalControllerIds = append(additionalControllerIds, controllerId)
+
+	if store.settingControllerListeners == nil {
+		store.settingControllerListeners = make(map[string]map[string][]ControllerSettingListener)
+	}
+
+	for _, id := range additionalControllerIds {
+		if store.settingControllerListeners[setting] == nil {
+			store.settingControllerListeners[setting] = make(map[string][]ControllerSettingListener)
+		}
+
+		if store.settingControllerListeners[setting][id] == nil {
+			store.settingControllerListeners[setting][id] = make([]ControllerSettingListener, 0)
+		}
+
+		store.settingControllerListeners[setting][id] = append(store.settingControllerListeners[setting][id], listener)
+	}
+}
+
```

Comment:
Not clear why you're deleting the listener if the controller specific settings are removed. Isn't the listener still going to want to react to changes to the global config?